Fetch token prices based on contract address

bitpshr opened the pull request:
This PR updates the Redux logic around token prices throughout the UI. It switches to using a Balanc3 API to query for token prices based on contract address instead of token symbol, fixing a bug that allowed arbitrary tokens sharing known symbols to appear incorrectly valued.

Additionally, parts of the UI were broken due to race conditions around token pricing. For example, the `TxListItem` component called `getSendTokenTotal` during its `componentDidMount` method and relied on data that was fetched by an unrelated `TokenCell` component. This never happened in time, and manifested itself by causing token transactions to never show a fiat value.

Lastly, this PR keeps token prices slightly more up-to-date with real-time exchange rates. Rather than querying for token data upon certain components' mounting and then never again, token prices are updated 1) anytime new app state is set, which could contain new tokens and 2) anytime tokens are added or removed during runtime.

Resolves #3981

bitpshr committed:
Fetch token prices based on contract address

danfinlay reviewed (changes requested):
Great approach, great bug fix, I think we need to rate limit the price checking though.

alextsg commented on ui/app/actions.js:
```diff
@@ -1080,9 +1082,12 @@ function unlockMetamask (account) {
+      type: actions.UPDATE_METAMASK_STATE,
+      value: newState,
+    })
+    dispatch(updateContractExchangeRates())
```
+1 on the rate limiting. `UPDATE_METAMASK_STATE` can happen pretty often and this would be expensive for a user with a big list of tokens.

bitpshr replied:
Ah! Had no idea it could be called so often. I will tweak the frequency for sure.

bitpshr committed:
Maintain token prices using a background service

bitpshr commented:
Thanks for the review @danfinlay and @alextsg. I added a `TokenRatesController` and removed all heavy-lifting from the user's browser context. The controller polls for token exchange rates on a configurable interval which defaults to 180 seconds. It uses an `ObservableStore` internally like other controllers so it can be composed as part of the `ComposableObservableStore` that makes up `metamask` Redux state.

I like this approach much better, thanks for the guidance.

kumavis commented:
These changes look good and match our current patterns.

Based on our research so far optimizing our outgoing http requests, I'm inclined to move anything that is polling over to the front end so that we're not polling for things we don't actively care about. 
some approaches:
- poll on the front end
- set up a subscription to the backend and poll when theres a subscription (+ automatically tear down subscriptions when the ui disconnects) -- we dont currently do anything like this
- accept it as technical debt (it solves active bugs) and merge as is

bitpshr committed:
Disable token price polling when no client is active

danfinlay reviewed (approved):
This looks clean, very nice, small change to add throttling when not in use!

bitpshr commented:
Thanks for the feedback @kumavis and @danfinlay. I worried about adding more and more asynchronous Redux actions to the already-heavy frontend and liked the idea of a background controller that continually kept token rates current. I ended up keeping the `TokenRatesController` in place but added measures to mitigate any potential DDOSing of Balanc3 APIs. Specifically, I added a reliable notion of both `isClientOpen` and `isClientOpenAndUnlocked` to the `MetaMaskController`, the latter of which is now used to set an `isActive` property on the rates controller that enables or disables its polling.

These extension points may also lend themselves to other work where controllers can safely be shut off during times of inactivity, such as #3971.

danfinlay commented:
One last thing you should start doing on your PRs, @bitpshr, is add a note to the `CHANGELOG.md` under `master`, which we add a new version heading to each time we publish a new version.

The entry should be user-facing, and describe the change as it affects end-users (including developers), and shouldn't focus on how it was implemented.

bitpshr committed:
Update CHANGELOG with information about rate fetching

bitpshr commented:
Thanks @danfinlay, added.

danfinlay committed:
Merge branch 'master' into i3981-contract-rates
